CONTRIBUTING.md

```diff
@@ -3,16 +3,53 @@
+* Functional tests are required to accompany any logic that modifies `pkg/apis/` API structs.
+* Follow the tide review flow for merging commits. This flow forces a review and all required CI jobs to pass before a PR is merged.
+* Read and understand the design principles outlined in this document that pertain to the code you are contributing to.
 
 ## Certificate of Origin
 
 By contributing to this project you agree to the Developer Certificate of
 Origin (DCO). This document was created by the Linux Kernel community and is a
 simple statement that you, as a contributor, have the legal right to make the
 contribution. See the [DCO](DCO) file for details.
+
+# Reconcile Loop Design Principles
+
+* Never block reconcile execution. ever. This means __no sleeps and no retry loops__
+* All __logic must be idempotent__. This means discovering what has already occurred by inspecting informer cache, and only mutating to converge on desired state.
+* __Never perform an apiserver List()__ request within the reconcile loop. This action is very apiserver intensive. Instead use informers retrieve items from cache.
+* __Avoid mutating a cluster object the loop didn't create__. A exception to this would to be adding annotations to an existing object. Mutating a object's Spec that our controller didn't create should be avoided.
+* __Avoid performing an apiserver Get()__ request within the reconcile loop if at all possible. Again use informers. There are rare exceptions.
+* Use __finalizers__ in order to perform any cleanup tasks required before a CR is completely removed from etcd.
+* User __owner references__ on any objects created and owned by a CR. Owner references ensure objects are automatically garbage collected after a corresponding CR is deleted.
+* Return early to handle errors which re-enqueues the work queue key to be tried again.
+* Do not create logic to sync a key using a recurring time interval. Instead only wake up processing the key when something has changed.
+* Reconcile execution should be broken into two parts, synchronization and status reporting. Synchronization converges the CR on the desired state. Status reporting records the current state of the CR based on the synchronization execution results and the observed state of the cluster. These are two independent functions. __Do not modify status during the synchronization step__.
+
+# API Design
+
+* All workflows __must be declarative__. This means the cluster-admin must be able to declare the exact state they want up front, and have our operator handle converging on that state.
+* __No imperative actions__. This means no requirement for a cluster-admin to account for the ordering or timing at which manifests are posted. If our API can't express the desired state up front, then the API is wrong.
+* Limit the Status.Phase enum to only express the state in broad terms. Use events and conditions on the CR to express detailed error information.
```

Comment:
I'm tempted to say "avoid phase at all, only use conditions".
https://github.com/kubernetes/kubernetes/issues/7856

Reply:
very interesting. I tend to agree after reading through that thread. I'll revise this a bit.

Reply:
revised, and thanks for the suggestion, i like this one.